Proposed changes for 1.0 (updated source repo)

acdha opened the pull request:
This is a replacement for #17 reflecting the move from the old loc-rdc organization to the primary LibraryOfCongress. The primary notable change from #17 is restoring the `fetch.txt` section following discussion with @jkunze, @dbrunton, and @johnscancella.

johnscancella committed:
first cut at bagit 1.0 spec

johnscancella committed:
updated ed summers, and justin littman addresses

acdha committed:
Remove legacy comment about old DNS results for xml.resource.org

acdha committed:
Trim trailing whitespace

acdha committed:
Remove long-commented “other file metadata” section

This appears to have been commented out since at least 2008.

acdha committed:
Strengthen guidance on UTF-8 BOMs in tag files

This follows the guidelines in RFC-3629

acdha committed:
Recommend modern hash algorithms

acdha committed:
Update recommendations related to Windows filenaming

* Note the existence of namespaces in the security considerations section
* Update previously un-displayed list of reserved DOS/Windows filenames

acdha committed:
Update interoperability disclaimer

Clarify that this section is part of the specification
but is not considered a hard requirement for an implementation.

acdha committed:
Discourage the use of manual bag creation

Update the section describing md5sum’s output format and clarify that it is strictly optional to accept bags which are produced using md5sum and will not pass a strict validation.

acdha committed:
Add filename-related normalization discussion

This adds background information for problems related to
case-sensitivity and Unicode normalization and adds a list of
recommendations for implementors.

acdha committed:
Update "Payload-Oxum" documentation

This adds the note that, unlike other metadata tags, this element must
not be repeated and clarifies that the Payload-Oxum value is not
sufficient for validation.

acdha committed:
Consistent number of spaces after a period

acdha committed:
Upgrade must/must-not text to entity references

This triggers the standard formatting in HTML, etc. outputs

acdha committed:
Strengthen wording: all manifests MUST list all files

acdha committed:
Better references syntax

* Use <organization> for relevant <author> entries
* Omit empty <date> attributes

acdha committed:
Update transfer recommendations

* Remove reference to GRABIT since the spec is now 
  returning HTTP 404 and there are no known public
  implementations.
* Add METALINK (RFC 5854) as an alternative which
  supports mirrors and protocols such as BitTorrent.

acdha committed:
Consistent indentation in Terminology

acdha committed:
Clarify the format for tag manifest algorithm names

johnscancella committed:
Cherry pick: first cut at changes from recommendations by Dave Crocker

acdha committed:
Update interoperability reference

acdha committed:
Update terminology for bag checksum algorithms

acdha committed:
Consistent indentation for terminology list

acdha committed:
Terminology: simplify “tag file” definition

acdha committed:
Terminology: simplify “valid” definition

This wording is shorter and doesn’t distinguish between
validation for payload and tag files.

acdha committed:
Convert numbered list in Structure section to <list>

acdha committed:
Remove Serialization section

The spec shouldn't need to include mechanistic transfer details: if the
results validate, it's a bag.

acdha committed:
Use <figure><artwork type="abnf"> for ABNF diagrams

acdha committed:
Minor <figure> consistency cleanup

acdha committed:
Terminology: whitespace cleanup

acdha committed:
Convert tabs to spaces

acdha committed:
Terminology: update definition of “complete”

johnscancella committed:
fixed some wording of sentences and removed section on Disk and network transfer since it doesn't really belong.

acdha committed:
Remove stale references 

The prose which referenced these was removed in 757f68ae9332944c771f0b3bd7491027b220e93d

acdha committed:
Remove trailing whitespace

acdha committed:
Remove stale inline TODO items

acdha committed:
Normalize XML declarations against current IETF template

* Update to follow https://www.rfc-editor.org/materials/template-bare-06.txt more closely
* Define <rfc> docName and category attributes

acdha committed:
Prose review for section 1

acdha committed:
Fix formatting for Section 2 (“Structure”)

acdha committed:
Fix formatting for “Bag Declaration”

acdha committed:
Copy-editing for “Payload Directory”

acdha committed:
Copy editing for “Payload Manifest”

acdha committed:
Remove injunction against reusing payload manifest algorithms

This appears to be redundant since the only way an algorithm could be reused would involve violating the
naming convention.

acdha committed:
Copy-editing for “Payload-Oxum”

acdha committed:
Simplify “Other Tag Files” text

acdha committed:
Spelling

acdha committed:
Copy-editing for “Complete, Incomplete, and Valid bags”

acdha committed:
Update version number in example bag

acdha committed:
Update authors

* Add Rosie Storey, David Brunton, Kate Zwaard
* Update email address for Justin Littman

acdha committed:
Clarify wording about directories in manifests

acdha committed:
Remove duplicate wording

acdha committed:
Update “filesystem” reference in “complete” requirements

Since people may wish to use things which aren’t commonly
considered filesystems (object stores, archive files, etc.)
this should be more generic.

acdha committed:
Update “special directory characters” prose

acdha committed:
Note that bag-info.txt fields are intended for human consumption

acdha committed:
Update Makefile and add a format target

acdha committed:
XML formatting

This applies consistent XML formatting going forward. It's a big diff
but does not change the generated text or HTML output.

acdha committed:
Remove `Bag-Size` from example bag-info.txt

This field is not specified but its presence in the examples can cause
confusion that it's a standard field. Since any app which needs a
human-formatted total file size may trivially calculate it from the
Payload-Oxum (also see #2) or filesystem metadata this commit removes it
to avoid confusion or the implication that tools must maintain the
value.

acdha committed:
Fix Makefile targets

acdha committed:
Tidy source formatting in the references section

acdha committed:
Add reference to conformance suite project

acdha committed:
Restore fetch.txt

We'll table discussion about the issue of transfer versus storage for BagIt 2.

acdha committed:
Add a reminder that fetch.txt entries must be listed in manifests

acdha committed:
Convert fetch.txt artwork to ABNF

acdha committed:
Remove dead comments

acdha committed:
Tidy format of manifest section

* Convert free-form list of statements to an actual <list>
* Indentation

acdha committed:
Update payload manifest to prohibit duplicates

* Specify that files are only listed once
* Convert list of requirements to an actual <list> so each
  point is notable

acdha committed:
Clarify that “whitespace” means tabs / spaces

We don't promise that every unicode code point flagged as
a space separator is valid.

acdha committed:
Stricter bagit.txt wording

This file isn't intended to be human edited and the cost of lax parsing
is not worthwhile.

acdha committed:
Tighten restrictions for bag-info.txt syntax

Previously the spec allowed variable whitespace before and after the
separator in bag-info.txt files. This commit changes that to a single
whitespace character for 1.0 to reduce implementation complexity and
reflect the widespread availability of validation tools.

johnscancella committed:
fixed ABNF for bag-info.txt and added one for bagit.txt

acdha committed:
Include Git configuration used for pretty diffs

The included .gitattributes and config change will diff the XML file by
running it through xml2rfc so you can see the visible text changes
rather than the XML source.

This requires one config change which cannot be included in the repo for
security reasons:

    git config --local diff.xml2rfc.textconv "xml2rfc --quiet --out=/dev/stdout"

If someone does care about the actual XML differences, the default
conversion can be disabled:

    git diff --no-textconv

Commit:
Changes in preparation for 1.0

Commit:
Requested tweeks.

johnscancella committed:
Merge pull request #1 from justinlittman/bagit1.0-changes

Bagit 1.0 changes

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+  </artwork>
+          </figure>
+          <t>
 A tag manifest file has the same form as the payload file manifest
```
Is the tag manifest also required to be complete and equivalent across hash algorithms?

acdha replied:
That was our intention and it's a good point for clarification

stain replied:
I've added https://github.com/LibraryOfCongress/bagit-spec/pull/19 to clarify that intention - left it as a "SHOULD list all tag files" unless you agree it would be MUST.

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+          <t>
+            <list style="symbols">
+              <t>The hex-encoded checksum &may; use uppercase and/or lowercase letters.</t>
+              <t>The slash character ('/') &must; be used as a path separator
```
filename (or should we say `filepath`?) SHOULD NOT contain `./` or `../` path segments. 
e.g. the line should (or MUST?) be 
```
b1946ac92492d2347c6235b4d2611184 data/file1.txt
33f4aa9ea510ffd5bd80f5fd85e40803 data/nested/file2.txt
```
and these should (must?) NOT be valid:
```
b1946ac92492d2347c6235b4d2611184 ./data/file1.txt
33f4aa9ea510ffd5bd80f5fd85e40803 data/nested/deeper/../file2.txt
```

acdha replied:
@johnscancella can correct me if I'm wrong but I think at least `./` is common enough that we'd be forcing 0.97 bags to be rewritten for 1.0. I've been thinking of this as a 2.0 thing where we could stop promising 100% compatibility with old md5deep output conventions.

johnscancella replied:
@acdha I believe that to be the case. For the java implementation I believe we will collapse down the filepath so that any `./` or `../` will be removed so moving to a new version should be fairly simple. But I agree this should probably wait until 2.0 since we will be doing breaking changes in that version anyway

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+                If a <spanx style="emph">filename</spanx> includes a newline
+                (LF), a carriage return (CR),
+                or carriage return plus newline (CRLF) it &must; be
+                percent-encoded following <xref target="RFC3986"/>.
```
This becomes a bit inconsistent as RFC3986 would escape other characters as well.. but only if there is CR/LF/CRLF?

```
33f4aa9ea510ffd5bd80f5fd85e40803 data/filename with spaces and 50% discount.txt
b1946ac92492d2347c6235b4d2611184 data/filename%20with%20newline%10and%20spaces%20and%2050%25%20discount.txt
```

stain replied:
.. in addition RFC3986 requires ASCII `%` escapes of any unicode characters, but we explicitly expect the unicode encoding here.

acdha replied:
This is an artifact of RFC 3986 being used in different (i.e. non-Unicode) contexts. I think the right way to read this is that we're using the RFC 3986 escaping format but only for the smaller set characters which have meaning in BagIt.

stain replied:
I attempted to addressed this in https://github.com/LibraryOfCongress/bagit-spec/pull/17

Note that even if we only escape CR/LF we must also escape `%` to `%25` as otherwise you don't know if %-encoding was used or not.  

(Alternatively we must say that the filepath before bagging can never contain `%0d` or `%0a`  - which begs the question of why to use %-escaping here..)

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+            The files under the payload directory are called payload files, or the payload.
+            Each payload file is treated as an opaque octet stream when verifying file
+            correctness.
+            Any sub-directory structure within the payload &must-not; be changed but is
```
I don't understand "must not be changed"?

acdha replied:
The intention of that section is that the directory hierarchy should be preserved but we don't otherwise care about it. Perhaps there's a better wording for that

stain replied:
Addressed in https://github.com/LibraryOfCongress/bagit-spec/pull/13

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+            metadata elements &must; be preserved.
+          </t>
+          <t>
+            A metadata element &must; consist of a label, a colon, at least one
```
Any reason why we can't simply build on [RFC2822](https://www.ietf.org/rfc/rfc2822.txt) for `bag-info.txt` fields? At least for line continuation and label definition?

johnscancella replied:
Because of historical reasons. This version is just to take what we already have out of draft and make it an official RFC. There will be another version coming soon afterwards that will have many backwards breaking updates.

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+              This element &must-not; be present more than once and, if present, &must;
+              be in the form "<spanx style="emph">OctetCount</spanx>.<spanx style="emph">StreamCount</spanx>",
+              where <spanx style="emph">OctetCount</spanx> is the total number of
+              octets (8-bit bytes) across all payload file content and
```
Does this include payload files referenced from fetch.txt or just the files who currently appear in the `data/` directory?

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+              A sender-supplied identifier for the bag.
+          </t>
+          <t hangText="Bag-Size">
+              Size or approximate size of the bag being transferred, followed
```
Presumably `Bag-Size` also include the size of the tagfiles and directory paths? (e.g. size of a uncompressed tar-ball or `du -h .`

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+              Date (YYYY-MM-DD) that the content was prepared for delivery.
+          </t>
+          <t hangText="External-Identifier">
+              A sender-supplied identifier for the bag.
```
Should `External-Identifier` be a URI? 

Should this be a `urn:uuid:` [UUID](https://tools.ietf.org/html/rfc4122)?

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+              should make an effort to honor reference to it.
+          </t>
+          <t hangText="Bag-Count">
+              Two numbers separated by "of", in particular, "N of T",
```
The `Bag-Group-Identifier` SHOULD be present if `Bag-Count` is present?

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+          </list>
+          </t>
+          <t>
+          In addition to these metadata elements, other arbitrary metadata
```
.. which should be valid according to RFC2822.

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+          <t>
+            The slash character ('/') &must; be used as a path separator in
+            <spanx style="emph">filename</spanx>. One or more linear whitespace
+            characters (spaces or tabs) &must; separate these
```
This means that `filename` can't begin with ` ` but that is OK because it must begin with `data/` :)

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+            a subset of logically related components (e.g., the localized logical
+            object could be much larger than what is intended for export) or
+            assembled from logically distributed sources (e.g., the object components
+            for export are not stored locally under one filesystem tree).
```
It should be clarified here if the total size of the unfilled files should be included in `Payload-Oxum` (I think it should).

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+            limitation on the length of any of the fields in the fetch file.
+          </t>
+
+          <t>
```
See comments in 
* #7 (semantics if file exists - assume it is also fetchable from that URI, but clients should validate remote checksum)
* #9 (can we have multiple/atlernate entries for same path? I hope yes - but they should have same checksum!)
* #10 (content-negotiation? Extra headers indented. Perhaps too drastic for 1.0? This is however time to do backwards-breaking changes)

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+
+          <t>
+            For reasons of efficiency, a bag &may; be sent with a list of files to be
+            fetched and added to the payload before it can meaningfully be checked
```
It is implied from above:
> Every payload manifest &must; list every payload file exactly once.

.. that these fetched files MUST also be listed in the payload manifest (so the bag can be checked as valid). Can this be made explicit here for guidance?

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+            specification.
+
+            Implementations &must; perform standard checksum validation on any tag file
+            which is listed in a tag manifest but &must; otherwise ignore their contents.
```
That makes those other tagfiles pointless. They MAY otherwise ignore their contents.

acdha replied:
That's a good point to clarify — I'm tempted to remove “but …” entirely but perhaps that should just be adding “for the purposes of bag validation”?

stain replied:
Agree to remove the whole "but" clause. They can do with the remaining tag-files as they like, even during validation. 

For instance if the BagIt profile is https://w3id.org/ro/bagit/profile then a conformant implementation may also want to check that `manifest/metadata.json` is present and valid according to JSON-LD specification.   (If that fails, the bag is still "valid" according to BagIt spec, but not from the RO Bagit profile).

We could add a similar "opaque" clause as for the payload files, but then it gets tricky because SOME tagfiles have meaning for this spec.

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+</t>
+      <t>
+The use of UTF-8 for text tag files is strongly &recommended;. Future version
+of BagIt may disallow encodings other than UTF-8.
```
This very spec version disallows other encodings because `bagit.txt` has a fixed line of
```http
Tag-File-Character-Encoding: UTF-8
```
Why leave this open?

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+manifest's filename by lowercasing the common name of the algorithm and
+removing all non-alphanumeric characters. Following is a partial list
+mapping common algorithm names to normalized names:
+          <list style="symbols">
```
Can we reference one of the existing [IANA registries](https://www.ietf.org/assignments/) here? If not then when moving to RFC we would effectively be starting our own registry.

https://www.ietf.org/assignments/named-information/named-information.txt from https://tools.ietf.org/html/rfc6920#section-9.4 is quite good (and adding to that registry is not tied to modifications to protocols) -- it deliberately does not list weak hashes MD5 and SHA1, but we can reference those two explicitly here.

A good question from there is what to do with truncated hashes which are becoming common - `sha-256-120` would become confusingly `sha256120` in the normalization proposed here. 

Also there are already some confusion between SHA2 512 and SHA3 512, presumably your `sha512` means `hmac-sha2-512`? again `sha3-512` would become `sha3512` in this normalization.

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+  algorithms <xref target="RFC6234"/> and &should; enable SHA-512 by default
+  when creating new bags.
+
+  For backwards-compatibility implementers &should; support
```
unclear if "supports should support" mean readers or writers. So should a new bag contain md5/sha1? I hope not.

stain commented on bagit.xml:
```diff
@@ -825,654 +806,432 @@ myfirstbag/
-     |    (... list of crawler starting point URLs ...                )
-     ....
+|   bag-info.txt
+|    (Source-URL: https://www.loc.gov/resource/highsm.23364/       )
```
This example should not be showcasing a non-standard field. While I like the idea of this particular property it is confusing compared to `External-Identifier` and this example.

johnscancella replied:
bag-info is at the heart just a list of key values fields. The reserved fields which you are talking about are optional and only have special meaning if the receiving party derives special meaning from a particular field.

stain replied:
This is suggesting `Source-URL` in an example - yet it is not defined. Examples of extensions should be shown separately. This example is meant to show a fetch.txt-driven bag.

stain replied:
Addressed in https://github.com/LibraryOfCongress/bagit-spec/pull/18

stain commented on bagit.xml:
```diff
@@ -825,654 +806,432 @@ myfirstbag/
+    </section>
+    <!-- /Examples -->
+    <section title="Security Considerations" anchor="sec-security">
+      <section title="Special directory characters">
```
Some of https://tools.ietf.org/id/draft-soilandreyes-arcp-03.html#rfc.section.7 might also apply.

stain commented on bagit.xml:
```diff
@@ -825,654 +806,432 @@ myfirstbag/
+          namespace sequences (e.g. "\\?\C:\&#x2026;") described in <xref target="MSFNAM"/>.
+        </t>
+        <t>
+          To assist implementers, the <xref target="LC-CONFORMANCE-SUITE">Library
```
This is marketing and should not be provided in RFC.

acdha replied:
I would respectfully disagree that a reference to an open-source test suite is just marketing. The name was included to avoid canonicalizing it as the “official” suite but we have used it to correct interoperability issues in several implementations so it's useful for anyone implementing the spec.

stain commented on bagit.xml:
```diff
@@ -825,654 +806,432 @@ myfirstbag/
+        <t>
+          Windows implementations &should; test their implementations to ensure
+          that safety-checks prevent use of drive letters and the less commonly used
+          namespace sequences (e.g. "\\?\C:\&#x2026;") described in <xref target="MSFNAM"/>.
```
As well as [Alternative Data Streams](https://blogs.technet.microsoft.com/askcore/2013/03/24/alternate-data-streams-in-ntfs/) which could be used to hide files (including executables) inside another file entry.

acdha replied:
That raises some other questions — that's in the Windows section but most other operating systems have similar concepts (extended attributes or resource forks) and there are non-trivial portability concerns trying to use them. 

Technically this could currently be considered a client implementation issue – e.g. a Windows bag creation tool could have manifests listing `foo.txt`, `foo.txt:$DATA`, etc. but that'd definitely increase the odds of someone experiencing compatibility issues moving that to other systems.

I'm inclined to say that BagIt should discourage all of this in favor of portability and have a suggestion that if you need to preserve non-portable filesystem metadata you create an archive using the platform's preferred format and bag that.

stain replied:
The Alternative Data Streams are commonly used for attacks such as hiding root kits. Several [CVEs](https://cwe.mitre.org/data/definitions/69.html) have been raised on this.

So if someone maliciously mints a bag with such "files" in it, when extracted they would disappear from regular listing in the filesystem (but can still be executed). 

Perhaps this is more of a problem for `fetch.txt` than the manifest - of course if the bag is in an archive this is a problem with a naive unarchiving process, the malicious user would not really need to list it in the manifest (unless the implementation is stringently checking against its listing during extraction) as most likely the "complete" check of the archive would not check for ADS-es.

acdha replied:
I think this is an implementation issue — and e.g. many places require virus scanning — but it might be useful to put a note about that along with the other points in the implementers guide since we already have other notes about security and filesystem quirks. Checking for alternate data streams seems like a good area to encourage awareness around since they're not uncommon.

stain commented on bagit.xml:
```diff
@@ -1,819 +1,800 @@
+            <artwork>url length filename</artwork>
+            <postamble>
+              where <spanx style="emph">url</spanx> identifies the file to be
+              fetched and must be an absolute URI as defined in
```
.. in addition RFC3986 requires ASCII `%` escapes of any unicode characters, while IRIs [RFC3987](https://www.ietf.org/rfc/rfc3987.txt) permits international characters, e.g. . We should be clear about which one we mean as we have UTF-8-mandated encoding also for fetch.txt.

E.g. which one we do require? 

I would prefer IRIs RFC3987:

```
http://résumé.example.org/café 1024 data/résumé/café.txt
```

vs the current URI requirement RFC3986:

```
http://xn--rsum-bpad.example.org/caf%C3%A9 1024 data/résumé/café.txt
```

(Note that IRI still permit `%`-style ASCII URIs):

stain reviewed (commented):
Some comments here - don't really think they any of them are blocking, although they might be good to clarify for 1.0

acdha committed:
Clarify that tag manifests should also list every manifest

This clarifies an issue raised by @stain:

https://github.com/jkunze/bagitspec/pull/19#discussion_r178067527

johnscancella committed:
refs #3 - updated ABNF to better match the normative

johnscancella committed:
refs #3 - added ABNF reference and not about using core rules

zimeon committed:
Metadata elements are normatively OPTIONAL

zimeon committed:
Be more specific than SHA-2 family

johnscancella committed:
Merge pull request #6 from zimeon/sha2_specifics

Be more specific than "SHA-2 family"

johnscancella committed:
Merge pull request #5 from zimeon/metadata_optional

Metadata elements are normatively optional

johnscancella committed:
refs #2 - added language to specify that all files must be in all manifests starting with version 1.0

johnscancella committed:
refs #7 - changed section about bags differing in normalization only from 'MUST' to 'SHOULD' be prevented

johnscancella committed:
refs #2 - fixing grammer

stain committed:
Tidy ABNF (e.g. shorter line length)

.. to avoid xml2rfc warnings

stain commented:
Don't forget to update  when merging:

```xml
    <date day="6" month="December" year="2016"/>
```

stain committed:
Can't say "only cryptographic-quality"

.. and then link to a section that also lists md5 and sha1

stain committed:
Use sha512 as example, not sha1

.. also no need to show normalization here, as we link to subsection

stain committed:
Payload is a set of named files

.. not a single stream of octets

(Should we have "payload files" in the terminology as well?)

stain committed:
List all types of tag files

.. not just bag-info and tagmanifest

stain commented:
It is unclear in section bagit.txt 

> The "bagit.txt" tag file MUST consist of exactly two lines in this order:
> ```
> BagIt-Version: M.N
> Tag-File-Character-Encoding: UTF-8
> ```
> M.N identifies the BagIt major (M) and minor (N) version numbers, 
> and UTF-8 identifies the character set encoding used by the tag files. 
> The bag declaration MUST be encoded in UTF-8, and MUST NOT 
> contain a byte-order mark (BOM) [RFC3629]. 

This can be read as that the bag MUST always have UTF-8 as `Tag-File-Character-Encoding` - which makes considerations in 2.3. Text Tag File Format moot.  

But on second reading do I understand that you still want to allow any encoding (without saying where that encoding name is defined) - and that it is the bagit.txt file itself that is the only one that must be UTF-8? (why not ASCII?)

I would not mind voting for fixed UTF-8 for Bagit 1.0. This has become the norm for most formats like XML, JSON. If we allow other character encoding we must say which registry we refer to, otherwise arbitrary encoding strings like "code page 865" would be allowed.

**Edit**: See https://github.com/LibraryOfCongress/bagit-spec/pull/14

stain committed:
*manifest-sha512* in examples

.. not md5/sha1 -- people might get the wrong idea!

stain committed:
MD5 has never been called "MD-5"

stain committed:
Explain why example use md5

sha512 would break our figure width!

stain committed:
Avoid confusing "sub-directory must not be changed"

It is out of scope for this specification to specify "changes" to bags
or the process of archiving an existing directory file structure into
bagit.

Rather we should say that the payload files may be in sub-directories
(data/ is not flat), and just that the directory and filenames have no
given meaning .. HERE.

Obviously, just like the octet bytes they have a
meaning to someone.

(If we are going to talk about "change" this opens a can of worms about
changes to checksums, changes to tag files, augmenting bag-info.txt etc
etc)

johnscancella committed:
Merge pull request #8 from stain/abnf-tidy

Tidy ABNF (e.g. shorter line length)

johnscancella committed:
refs #2 - changed text to be more clear that the previous version is just a note

johnscancella committed:
updated date

johnscancella committed:
changed the bagit.txt example to make it more clear that you can choose which encoding the other tag files are in

stain committed:
Typo "ather" -> "rather"

Fixes
https://github.com/LibraryOfCongress/bagit-spec/pull/10#pullrequestreview-114182845

stain committed:
filenames, not file names

Fixes
https://github.com/LibraryOfCongress/bagit-spec/pull/13#pullrequestreview-114182762

stain committed:
files may be organized in subdirectories

stain committed:
correct plural (many files, many sequences)

stain committed:
Avoid RFC "MAY" for "base directory"

... as it implies you should only use "any name" if you really know what
you are doing. But really we don't recommend any particular name, so we
can't use MAY here.

stain committed:
Use "filepath" instead of "filename"

.. to imply that the filepath also will include relative directories.

Filename is often seen as just the last segment of a filepath.
"pathname" is on the otherside only the "directory" side.

filepath: data/fred/soup.txt
filename: soup.txt
pathname: data/fred/

stain committed:
Avoid unknown Source-URL in example

If we want to show "custom" keys then that should be a different
example

stain committed:
Attempt to clarify whitespace in bag-info.txt

The more liberal <1.0 format is not part of this spec,
and only mentioned as a SHOULD below.

Note that this still allows space WITHIN the label (but not start or
end). Example:

Valid:

Long Description Label: And even
  longer description value

Invalid (in 1.0):

Can't end with space : Can we?

 Cantstartwithspace: even on first line

stain committed:
About multiplicity of metadata elements

Some of these don't make sense to repeat, but except for
Payload-Oxum there is probably no need to be strict.

johnscancella committed:
Merge pull request #13 from stain/payload-directory-structure

Avoid confusing "sub-directory must not be changed"

johnscancella committed:
Merge pull request #10 from stain/sha512-example

Use sha512 as example, not sha1

johnscancella committed:
Merge pull request #11 from stain/payload-is-files

Payload is a set of named files

johnscancella committed:
Merge pull request #12 from stain/terminology-tagfile

List all types of tag files

johnscancella committed:
clarified that the encoding is used by the remaining tag files

johnscancella committed:
Merge pull request #15 from stain/confusing-MAY-clause

Avoid confusing MAY clause

johnscancella committed:
Merge pull request #18 from stain/avoid-source-url

Avoid unknown Source-URL in example

johnscancella committed:
Merge pull request #20 from stain/whitespace-ignored

Attempt to clarify whitespace in bag-info.txt

johnscancella committed:
Merge pull request #9 from stain/cryptographic

Can't say "only cryptographic-quality"

johnscancella committed:
added Stian Soiland-Reyes to list of contributors

stain committed:
RECOMMENDED instead of SHOULD BagCount/-GroupID link

.. as this is more of a best practice recommendation

stain committed:
Consistent "line feed (LF)" instead of "newline"

As agreed in
https://github.com/LibraryOfCongress/bagit-spec/pull/17#discussion_r184976972

stain committed:
Typo fiilepath

Fixes
https://github.com/LibraryOfCongress/bagit-spec/pull/16#discussion_r183479826

stain committed:
Reference RFC2989 registry for ENCODING

text as suggested by @acdha in
https://github.com/LibraryOfCongress/bagit-spec/pull/14#issuecomment-385392299

I also used <spanx> to separate ENCODING/MUST/UTF-8 as it got a bit
capsy :)

johnscancella committed:
Merge pull request #16 from stain/filepath

Use "filepath" instead of "filename"

johnscancella committed:
Merge pull request #21 from stain/unique-metadata-elements

About multiplicity of metadata elements

stain committed:
Merge branch 'bagit1.0' into line-feed-not-newline

johnscancella committed:
Merge pull request #27 from stain/rfc2978

Reference RFC2989 registry for ENCODING

johnscancella committed:
Merge pull request #26 from stain/line-feed-not-newline

Consistent "line feed (LF)" instead of "newline"

johnscancella committed:
refs #28 - changed wording for pull request

johnscancella committed:
refs #17 - fixed merge conflicts for only percent encode CR, LF, and CRLF

johnscancella committed:
refs #22 - fixed merge conflict for tag files have optional newline at end of file

johnscancella committed:
refs #23 - fixing merge conflict on using registry for hash names

johnscancella committed:
refs #19 - tag manifest should be complete and have the same list of tag files

Commit:
Tweaks.

Commit:
Fixed keywords in non-normative section.

johnscancella committed:
Merge pull request #29 from justinlittman/tweaks

Tweaks.

johnscancella committed:
Merge branch 'bagit1.0' into shoulds

johnscancella committed:
Merge pull request #30 from justinlittman/shoulds

Fixed keywords in non-normative section.

johnscancella committed:
updated date

jkunze committed:
increment draft number

jkunze committed:
tweak to author order

jkunze committed:
grouped unaffiliated authors together

jkunze committed:
added '-' affiliation to fix formatting glitch

An empty affiliation outputs a blank line after all but the last unaffiliated author in a group.

Commit:
refs #31. Makes changes requested from ISE review.

acdha committed:
Merge pull request #32 from justinlittman/ise_tweaks

refs #31. Makes changes requested from ISE review.

jkunze committed:
alphabetize contributors; fix typo

jkunze committed:
add two non-normative sentences to address ISE review comment about error handling

"Upon discovering errors in bags, an implementation is free to take action (for example, logging or reporting) in an application-specific manner. This document does not mandate any particular action."

jkunze committed:
added the missing noun ("rules") that the term "ABNF" qualifies

jkunze committed:
made artwork displays all end consistently

some displays ended with an extra blank line

jkunze committed:
reduced redundant use of "optional"

Per reviewer comment:
>    Section 2.1.3), a file named "bagit.txt" (see Section 2.1.1), and
>    zero or more additional tag files (see Section 2.2).  The tag files
>    in the optional tag directories are arbitrary file hierarchies and
>    the tag directories MAY have any name that is not reserved for a file
>    or directory in this specification.

Above (2) seems to say that all tag directories are optional.  Hence
constantly including the word 'optional' for them, in the rest of the
document, is distracting.

>
>    The base directory MAY have any name.
>
>            <base directory>/
>            |   bagit.txt
>            |   manifest-<algorithm>.txt
>            |   [optional additional tag files]
>            \--- data/
>                  |   [payload files]
>            \--- [optional tag directories]/
>                  |   [optional tag files]

The square brackets are probably enough to indicate being optional. The
word just makes things wordier.

_The word “optional” has been removed as redundant, given the bracketing and that all tag directories have been described previously as optional._

acdha commented on bagit.xml:
```diff
@@ -287,8 +287,7 @@ The base directory can have any name.
             |
-            +-- [optional tag files]
-  </artwork>
+            +-- [optional tag files] </artwork>
```
Was the intention of this to avoid an extra line in the rendered text output? I'm not a huge fan of the closing tag being on the end of the line like this but I'm not sure it's worth changing everything to go the other way.

jkunze replied:
Exactly, and it definitely makes the XML ugly. It may be a defect in the xml2rfc tool. If you can find a way around it, go for it, but if push comes to shove, I think getting the rendered human-oriented document consistent and correct is more important than making the XML pretty.

jkunze committed:
tightened "file" reference to "file name"

Per reviewer comment:
>    A payload manifest is a tag file that lists payload files and

probably:

    that lists payload file names and

 Clarified.

Saying "lists" does imply names and not the file contents, but for some
reason I think the modified form will be clearer.


>    checksums for those payload files generated using a particular bag

I'm pretty sure it's not the payload files that are generated using a
checksum algorithm...  I assume it's a manifest payload file listing...

_That sentence was stricken during recent editing rounds. A similar sentence has been reworded: “Every payload manifest MUST list every payload file name exactly once.”_

jkunze committed:
added an extra summary section sentence

Per reviewer comment:
>    checksum algorithm.  Every bag MUST contain one payload manifest
>    file, and MAY contain more than one.  A payload manifest file MUST

I think this is unusual enough to warrant, again, an initial, summary
statement.  If I'm understanding, it should be something like:

      A bag can have more than one data integrity manifest, with each
using a different validation algorithm.

_This sentence has been added: A bag can have more than one payload manifest, with each
using a different validation algorithm._

jkunze committed:
clarified metadata

Per reviewer comment:
>    Source-Organization  Organization transferring the content.
...
>    Organization-Address  Mailing address of the organization.

    organization  ->  source organization

>    Contact-Name  Person at the source organization who is responsible
>       for the content transfer.
>
>    Contact-Phone  International format telephone number of person or
>       position responsible.
>
>    Contact-Email  Fully qualified email address of person or position
>       responsible.
> ...
>    External-Description  A brief explanation of the contents and
>       provenance.
...
>    Bagging-Date  Date (YYYY-MM-DD) that the content was prepared for
>       delivery.

I think you mean 'transfer' rather than 'delivery'...

jkunze committed:
moved summary paragraph from end of section to beginning of section

Per reviewer comment:
>    The "fetch.txt" file allows a bag to be transmitted with "holes" in
>    it, which can be practical for several reasons.  For example, it
>    obviates the need for the sender to stage a large serialized copy of
>    the content while the bag is transferred to the receiver.  Also, this
>    method allows a sender to construct a bag from components that are
>    either a subset of logically related components (e.g., the localized
>    logical object could be much larger than what is intended for export)
>    or assembled from logically distributed sources (e.g., the object
>    components for export are not stored locally under one filesystem
>    tree).

This paragraph would be a better introduction to the section.

_Done._

jkunze committed:
clarified security implications of older checksum algorithms

Per reviewer comment:
>    Implementors of tools that complete bags by retrieving URLs listed in
>    a "fetch.txt" file need to be aware that some of those URLs may point
>    to hosts, intentionally or unintentionally, that are not under
>    control of the bag's sender.  Checksums are intended as a reasonable
>    guarantee against corruption during transit, not a strong
>    cryptographic protection against intentional spoofing.

Oh?

_This wording was meant to apply to checksums as they are used in bags, as well as to address criticism that many legacy bags used easily broken MD5 checksums. That last sentence has now been reworded to: Moreover, older checksum algorithms, even if reasonable for detecting corruption during transit, may not offer strong cryptographic protection against intentional spoofing._

jkunze committed:
usage tweak

Per reviewer comment:
>    In all text tag files except for the bag declaration file, text MUST
>    be encoded in the character encoding specified in the "bagit.txt" bag

    be encoded in the character encoding  -> use the character encoding

_Done._

jkunze committed:
beefed up response to security case

Per reviewer comment:
>    The size of files, as optionally reported in the "fetch.txt" file,
>    cannot be guaranteed to match the actual file size to be downloaded.
>    Implementors SHOULD take care to appropriately handle cases where the
>    actual file size does not match the file size reported in the
>    fetch.txt.  Implementors SHOULD NOT use the file size in the
>    "fetch.txt" file for critical resource allocation, such as buffer
>    sizing or storage requisitioning.

Absent specification of what "appropriately handle" means, this guidance
lacks substance.

_Reworded the second sentence to be: Implementers SHOULD take steps to monitor and abort transfer when the received file size exceeds the file size reported in the fetch file._

jkunze committed:
changed a MUST to a standard &must;

jkunze committed:
more consistent terminology (validation -> checksum)

jkunze committed:
bumped draft number and current date

justinlittman committed:
Update Justin's contact info

acdha committed:
Merge pull request #34 from justinlittman/patch-1

Update Justin's contact info

jscancella committed:
updated email address

justinlittman committed:
Added clarification about malicious attackers.

justinlittman committed:
Changed reference to character set registry.

acdha committed:
Merge pull request #35 from jscancella/contactInfoUpdate

updated email address

acdha committed:
Merge pull request #37 from justinlittman/patch-2

Changed reference to character set registry.

justinlittman committed:
Updates of wording.

acdha committed:
Merge pull request #36 from justinlittman/patch-1

Added clarification about malicious attackers.

jkunze committed:
Update bagit.xml

jkunze committed:
update draft number and draft date

acdha committed:
Final XML from IETF
